### PR TITLE
UrlSync: Support browser history steps, remove singleton

### PR DESCRIPTION
**Pull request by torkelo**
* Adds new function to SceneObjectUrlSyncHandler named shouldCreateHistoryStep that enable scene apps to control when new history steps should be created due to state changes
* Adds new feature to UrlSyncManager to sync initial scene state to URL, this is important as we need this initial URL state to go back to after a say a time range state
* remove getUrlSyncManager and remove the singleton nature of UrlSyncManager (needed for other reasons https://github.com/grafana/scenes/pull/841) 

To not make this a big new behavior I am adding new options to control it
* updateUrlOnInit 
* createBrowserHistorySteps (only SceneTimeRange is taking advantage of this)

can be set on SceneApp or UrlSyncContextProvider / useUrlSync 

Question is should they be inverted and so the default state is enabled (ie updateUrlOnInit => disableUpdateUrlonInit,   createBrowserHistorySteps => disableBrowserHistorySteps), so it's more opt-out than opt-in 

# Release notes

getUrlSyncManager is no longer exported as UrlSyncManager is now no longer global singleton but local to the UrlSyncContextProvider. 
If you called getUrlSyncManager().getUrlState that util function is available via the exported object sceneUtils. 

<details>
  <summary>📦 Published PR as canary version: <code>5.13.0--canary.878.10702174754.0</code></summary>
  <br />

  :sparkles: Test out this PR locally via:
  
  ```bash
  npm install @grafana/scenes-react@5.13.0--canary.878.10702174754.0
  npm install @grafana/scenes@5.13.0--canary.878.10702174754.0
  # or 
  yarn add @grafana/scenes-react@5.13.0--canary.878.10702174754.0
  yarn add @grafana/scenes@5.13.0--canary.878.10702174754.0
  ```
</details>

### CODE REVIEW

File: packages/scenes/src/services/UrlSyncManager.ts [hunk 1]
```diff
@@ -88,33 +125,36 @@ export class UrlSyncManager implements UrlSyncManagerLike {
...
+      if (!isUrlValueEqual(currentUrlValue, newUrlValue)) {
+        mappedUpdated[uniqueKey] = newUrlValue;
       }
+    }
 
-      if (Object.keys(mappedUpdated).length > 0) {
-        writeSceneLog('UrlSyncManager', 'onStateChange updating URL');
-        locationService.partial(mappedUpdated, true);
+    if (Object.keys(mappedUpdated).length > 0) {
+      const shouldCreateHistoryEntry = changedObject.urlSync.shouldCreateHistoryStep?.(newUrlState);
```
[the remark this answers is not in the document]
Comment (torkelo) [reply packages/scenes/src/services/UrlSyncManager.ts:148]: > What about changes that happen in a batch or are chained? Coming back to the example you brought in https://github.com/grafana/scenes/pull/760

It's tricky, the interesting thing to consider is when you do a new history your basically creating a step you can go back to. So time ranges changes (when initiated by the user at least) should be good "states". 

> What if we needed to create a history step but only upon a user action, e.g.:

You should be able to control this by control when shouldCreateHistoryStep return true (say right after a user action, but if it's a programmatic action you maybe return false). 

> user selects a datacenter -> new history is created
> cluster is changed automatically -> no new history is created, just updates the recent one
> job is changed automatically -> no new history is created, just updates the recent one

I have not implemented this for variables yet but my current thinking is to only create history items for calls to changeValueTo (as this is what the UI select calls), but when a variable changes value due to a parent / options change the this does not go through changeValueTo. So to handle this there would be some temporary state (next call to to shouldCreateHistoryStep should return true inside MultiValueVariable)

[the remark this answers is not in the document]
Comment (ifrost) [reply packages/scenes/src/services/UrlSyncManager.ts:148]: > You should be able to control this by control when shouldCreateHistoryStep return true (say right after a user action, but if it's a programmatic action you maybe return false).

The only argument that is passed to `shouldCreateHistoryStep` is the result of `getUrlState()`. How could we differentiate user vs programmatic actions based on this? 🤔 

>  but when a variable changes value due to a parent / options change the this does not go through changeValueTo.

Sounds interesting! Do you think we could have something similar not only for variables but also for setting state of a scene object in general?

[the remark this answers is not in the document]
Comment (torkelo) [reply packages/scenes/src/services/UrlSyncManager.ts:148]: @ifrost hm.. looks like GroupByVariable uses changeValueTo in updateFromUrl

so maybe we need a new parameter to changeValueTo(value, text, isUserAction = false)

[the remark this answers is not in the document]
Comment (torkelo) [reply packages/scenes/src/services/UrlSyncManager.ts:148]: did a quick POC here, https://github.com/grafana/scenes/pull/882 works really well

[the remark this answers is not in the document]
Comment (ifrost) [reply packages/scenes/src/services/UrlSyncManager.ts:148]: I tried to play with it a bit and works really nice! 🙌 

What about the case where a non-variable state is synced with the URL and should be pushed to history. Would consumers just would write customer wrapping like below?

```typescript
  public setFoo(value: number) {
    this._urlSync.performBrowserHistoryAction(() => {
      this.setState({ foo: value })
    })
  }
```

[the remark this answers is not in the document]
Comment (torkelo) [reply packages/scenes/src/services/UrlSyncManager.ts:148]: @ifrost yes, something like that

File: packages/scenes-react/src/contexts/SceneContextProvider.tsx [hunk 2]
```diff
@@ -79,5 +72,9 @@ export function SceneContextProvider({ children, timeRange, withQueryController
   }
 
   // For root context we wrap the provider in a UrlSyncWrapper that handles the hook that updates state on location changes
-  return <UrlSyncContextProvider scene={childContext}>{innerProvider}</UrlSyncContextProvider>;
+  return (
+    <UrlSyncContextProvider scene={childContext} updateUrlOnInit={true} createBrowserHistorySteps={true}>
```
Comment (ivanortegaalba) [inline packages/scenes-react/src/contexts/SceneContextProvider.tsx:76]: Should we allow to config `updateUrlOnInit={true} createBrowserHistorySteps={true}` here? Or the plugin dev should use a different `UrlSyncContext` provided to override those?

**Review by ivanortegaalba — COMMENTED** [review]
Took me a while to familiarize myself with all URL state management. The changes look good and the API is familiar and easy to use. I just would add some documentation and examples to the scenes doc, but we can leave it for a follow-up PR.

✅ The time range changes add a new history entry
✅ The initial time range doesn't add an extra history entry
✅ Resolving variables A -> B, when B is resolved, A state update doesn't add extra history entry
❌ The default dashboard time range is not updated when going back in the browser. Steps to reproduce in the video

https://github.com/user-attachments/assets/cdf3243f-8fef-4bbf-8e66-746751acb7db

**Review by ivanortegaalba — APPROVED** [review]

File: packages/scenes-react/src/contexts/SceneContextProvider.tsx (hunk 2 above)
Comment (torkelo) [reply packages/scenes-react/src/contexts/SceneContextProvider.tsx:76]: @ivanortegaalba well scenes react is kind still not used for real anywhere, so figured we can leave these enabled by default until for now

**Comment by torkelo** [comment]
@ivanortegaalba pushed some basic docs, we had zero docs on url syncing :)

**Pull request merged**